### PR TITLE
Support for Default HTTP Handler

### PR DESCRIPTION
**Pull request by sumeetchhetri**
This adds support where external framework integrations are needed.
If i need to integrate a new web framework, I can use seastar httpd as the base and provide the framework functionality on top of it.
Any matching rules will always over-ride the default handler.

### CODE REVIEW

File: include/seastar/http/routes.hh [hunk 1]
```diff
@@ -120,6 +120,14 @@ public:
      */
     routes& add(operation_type type, const url& url, handler_base* handler);
 
+    /**
+     * Add a global handler:
+     * Example  routes.add_global_hanlder(handler);
+     * @param handler
+     * @return
+     */
+    routes& add_global_handler(handler_base* handler);
```
Comment (nyh) [inline include/seastar/http/routes.hh:129]: Would be good to clarify what a "global" handler mean. I guess that this handler will get any HTTP method (unfortunately a new name "operation_type" was invented for that in the code) and URL? Maybe the word default, instead of global, is better, if you allow specific routes to be overridden in addition to that default (but currently you don't - more on this below).

File: src/http/routes.cc [hunk 2]
```diff
@@ -81,8 +81,11 @@ std::unique_ptr<reply> routes::exception_reply(std::exception_ptr eptr) {
 }
 
 future<std::unique_ptr<reply> > routes::handle(const sstring& path, std::unique_ptr<request> req, std::unique_ptr<reply> rep) {
-    handler_base* handler = get_handler(str2type(req->_method),
-            normalize_url(path), req->param);
+    handler_base* handler = _global_handler;
+    if(handler == nullptr) {
```
Comment (nyh) [inline src/http/routes.cc:85]: I think it's strange that in addition to the "global handler" you **allow** to set per-method and URL handlers, and yet ignore them if the global handler is set. Wouldn't it make more sense to have a "default handler", and allow other routes to override it?

**Review by nyh — CHANGES_REQUESTED** [review]
I didn't understand you you needed to change sstring's implementation to add this HTTP server feature. Seems unrelated unless I'm missing something. But even if the sstring change is really needed, please split this patch into two - one adding the sstring feature, and one adding the HTTP server feature.

The thing is, "sstring" is a very basic type in Seastar. Changing it will require good justification and motivation - more than the few words that you wrote. Please explain why this is an important feature, how also any downsides it might have (does it affects the "sstring" object length?). It will affect every Seastar user, so will require careful consideration. It can't be a part of a patch that adds a minor feature to the HTTP server.

File: include/seastar/http/routes.hh (hunk 1 above)
Comment (amnonh) [reply include/seastar/http/routes.hh:129]: We do support general path definitions (it's used for serving files for example) my question is there a need for a general operation?

File: src/http/routes.cc (hunk 2 above)
Comment (amnonh) [reply src/http/routes.cc:85]: The existing logic is: first check for exact URL match then check for matching-rules.

maybe you want to do it here

**Comment by sumeetchhetri** [comment]
> I didn't understand you you needed to change sstring's implementation to add this HTTP server feature. Seems unrelated unless I'm missing something. But even if the sstring change is really needed, please split this patch into two - one adding the sstring feature, and one adding the HTTP server feature.
> 
> The thing is, "sstring" is a very basic type in Seastar. Changing it will require good justification and motivation - more than the few words that you wrote. Please explain why this is an important feature, how also any downsides it might have (does it affects the "sstring" object length?). It will affect every Seastar user, so will require careful consideration. It can't be a part of a patch that adds a minor feature to the HTTP server.

Am creating a separate PR for the sstring changes

File: include/seastar/http/routes.hh (hunk 1 above)
Comment (sumeetchhetri) [reply include/seastar/http/routes.hh]: Subject line and comments for the PR updated with explanation and code updated based on the review comments

**Comment by sumeetchhetri** [comment]
Code updated as requested.

File: src/http/routes.cc [hunk 3]
```diff
@@ -81,8 +81,11 @@ std::unique_ptr<reply> routes::exception_reply(std::exception_ptr eptr) {
 }
 
 future<std::unique_ptr<reply> > routes::handle(const sstring& path, std::unique_ptr<request> req, std::unique_ptr<reply> rep) {
-    handler_base* handler = get_handler(str2type(req->_method),
-            normalize_url(path), req->param);
+    handler_base* handler = _default_handler;
+    handler_base* rule_handler = get_handler(str2type(req->_method), normalize_url(path), req->param);
+    if(rule_handler != nullptr) {
+        handler = rule_handler;
+    }
```
Comment (nyh) [inline src/http/routes.cc:88]: The logic of this change is unnecessarily complicated... You can leave the original `handler = get_handler()` code untouched, and just after it add the lines
```c++
if (handler == nullptr) {
    handler = _default_handler
}
```
But even better, the get_handler() function should just handle the default! In the end of the get_handler() function definition just return _default_handler instead of nullptr. That's it!

File: src/http/routes.cc [hunk 4]
```diff
@@ -140,6 +143,13 @@ routes& routes::add(operation_type type, const url& url,
     return add(rule, type);
 }
 
+routes& routes::add_default_handler(handler_base* handler) {
+    if(_default_handler == nullptr) {
+        _default_handler = handler;
+    }
```
Comment (nyh) [inline src/http/routes.cc:149]: I don't understand the if... Why not just set _default_handler without the if?
If you, for some reason, want to forbid overriding a previously-set default, please throw an exception. Don't just silently ignore the call.

**Review by nyh — CHANGES_REQUESTED** [review]

File: src/http/routes.cc (hunk 3 above)
Comment (sumeetchhetri) [reply src/http/routes.cc:88]: Good suggestion.

File: src/http/routes.cc (hunk 4 above)
Comment (sumeetchhetri) [reply src/http/routes.cc:149]: The exception will cause another thread that wants the same handler to be used to throw an exception, hence the if block, do let me know, I can remove the if clause though without any unwanted side-effects.

File: src/http/routes.cc [hunk 5]
```diff
@@ -81,8 +81,7 @@ std::unique_ptr<reply> routes::exception_reply(std::exception_ptr eptr) {
 }
 
 future<std::unique_ptr<reply> > routes::handle(const sstring& path, std::unique_ptr<request> req, std::unique_ptr<reply> rep) {
-    handler_base* handler = get_handler(str2type(req->_method),
-            normalize_url(path), req->param);
+    handler_base* rule_handler = get_handler(str2type(req->_method), normalize_url(path), req->param);
```
Comment (nyh) [inline src/http/routes.cc:84]: Please avoid unnecessary whitespace changes in the patch.

**Review by nyh — APPROVED** [review]
Other than the whitespace change (please avoid it), this version looks good to me. @amnonh please review this version again. Thanks.

File: src/http/routes.cc [hunk 6]
```diff
@@ -127,7 +126,7 @@ handler_base* routes::get_handler(operation_type type, const sstring& url,
         }
         params.clear();
     }
-    return nullptr;
+    return _default_handler;
```
Comment (amnonh) [inline src/http/routes.cc:129]: please note, that you will not be protected for mandatory parameters as other handlers do.

I don't know if that's an issue, but know that it's a limitation of your current implementation.

**Comment by amnonh** [comment]
Besides the question of mandatory parameters, which I'm not sure if it's by mistake or design, the rest LGTM

File: src/http/routes.cc (hunk 6 above)
Comment (sumeetchhetri) [reply src/http/routes.cc:129]: Thats exactly what should be provided by the default handler, a completely flexible interface

**Comment by sumeetchhetri** [comment]
Can this be merged @amnonh ?

File: src/http/routes.cc [hunk 7]
```diff
@@ -140,6 +140,11 @@ routes& routes::add(operation_type type, const url& url,
     return add(rule, type);
 }
 
+routes& routes::add_default_handler(handler_base* handler) {
+    _default_handler = handler;
```
Comment (amnonh) [inline src/http/routes.cc:144]: who will delete this handler?

Comment (sumeetchhetri) [reply src/http/routes.cc:144]: The calling application, as the application can choose to pass an object from the stack as the handler, I think this responsibility should not be seastar's.

Comment (amnonh) [reply src/http/routes.cc:144]: ok, just make sure you run it in debug mode and see that you don't get errors in the shutdown process, it seems like a potential problem

Comment (sumeetchhetri) [reply src/http/routes.cc:144]: Have validated locally, good to go from my side.

Comment (nyh) [reply src/http/routes.cc:144]: @amnonh this is a pre-existing "feature" in `add()` and `put()` as well - the caller needs to "new" a handler object, pass it by pointer, and it will never get freed. This is indeed very un-C++-like. Something like std::function<> would have probably made more sense.  But it makes sense for the new function to take the same "handler" type as the existing functions.

**Comment by nyh** [comment]
@amnonh what is your verdict on this patch? Shall I commit it?

Also, @amnonh, I have a question for you. I was worried that maybe since the same handler will be called for all HTTP methods (GET, POST, etc.) the handler will not know which method was used - because it's not one of the handler_base parameters. But looking deeper into the code I noticed that there is `request._method`, so that's fine. But then I started thinking - why does handler_base take a **path**? Why isn't that part of the `request` too?

**Comment by amnonh** [comment]
@nyh you can commit it.

The path parsed from the request and used inside the route object, I guess it's possible to add that functionality inside the request (even if the route object will do it)

File: src/http/routes.cc [hunk 8]
```diff
@@ -140,6 +140,11 @@ routes& routes::add(operation_type type, const url& url,
     return add(rule, type);
 }
 
+routes& routes::add_default_handler(handler_base* handler) {
```
Comment (psarna) [inline src/http/routes.cc:143]: nitpick: I think `set_default_handler` would be a better name. `add` suggests that you can add more than 1, while only a single instance is stored. Also, removing a default handler would be done with `add_default_handler(nullptr)`, which looks a little weird.

**Review by psarna — COMMENTED** [review]
Looks good, I'll commit in a minute. It would be good to follow up with a test, but the change is trivial enough for me to believe that it works.

File: src/http/routes.cc (hunk 8 above)
Comment (psarna) [reply src/http/routes.cc:143]: still, not a blocker at all, I'll commit the change. Thanks for the contribution!

**Comment by psarna** [comment]
Merged as 569433abcbc89a0bbd9595111e6a4e4ffbf0af12
